cpp/tensorrt_llm/kernels/trtllmGenKernels/blockScaleMoe/trtllmGenSrc/RoutingKernel.cu

```
@@ -0,0 +1,1340 @@
+    // implement break with EXIT.
+    __cluster_barrier_wait();
+
+    // trigger the secondary kernel when using PDL
+    // We can't do it earlier because FC1 depends on the mPtrCtaIdxXyToBatchIdx,
+    // mPtrCtaIdxXyToMnLimit, mPtrNumNonExitingCtas and mPtrTotalNumPaddedTokens
+    // TODO: this is not sufficient to ensure visibility in the next kernel!
+
+    // TODO: disable PDL for now to avoid race condition in FC1
+    if constexpr (KernelParams::UsePdl)
+    {
+        // cudaTriggerProgrammaticLaunchCompletion();
+    }
+
+// each thread has the same "expanded indexes" assigned to it as above
+// at this point, we know the final offsets of experts and the offsets within
+// experts, which allows writing the final index values
+#pragma unroll
+    for (int32_t ii = 0; ii < MaxExpandedIdxPerThread; ++ii)
+    {
+        auto expandedIdx = static_cast<int32_t>(clusterThreadIdx) + ii * NumThreadsPerCluster;
+        if (expandedIdx >= expandedIdxSize)
+        {
+            break;
+        }
+        auto expertIdx = expertIndexes[ii];
+        // check whether this expert is local to our GPU at all
+        auto localExpertIdx = static_cast<int32_t>(expertIdx) - params.mLocalExpertsStartIdx;
+        auto isLocalExpert = localExpertIdx >= 0 && localExpertIdx < localExpertExtent
+            && (localExpertIdx & params.mLocalExpertsStrideLog2) == 0;
+        auto tokenIdx = expandedIdx / NumTopExperts;
+        auto permutedIdx = isLocalExpert ? int32_t{smemExpertOffset[expertIdx]} + expertOffsets[ii] : int32_t{-1};
+        if (params.mPtrExpandedIdxToPermutedIdx != nullptr)
+        {
+            params.mPtrExpandedIdxToPermutedIdx[expandedIdx] = permutedIdx;
+        }
+        if (params.mPtrPermutedIdxToTokenIdx != nullptr && isLocalExpert)
+        {
+            params.mPtrPermutedIdxToTokenIdx[permutedIdx] = tokenIdx;
+        }
+    }
+}
+#else
+__global__ void routingIndicesClusterKernel(KernelParams params)
+{
```

Need to throw an exception here.

---

Sure, I'll do it in the another PR.